Use better cd_values[] in LZ4 example

ajelenak opened the pull request:
The old value does not seem appropriate for the LZ4 filter as it is too small. Such small values will create much larger compressed chunks than the original chunk size.

byrnHDF commented:
Right so you need to update the testfiles references to match the output of the tests

ajelenak commented:
It seems more than just that. The example fails on this line: https://github.com/HDFGroup/hdf5_plugins/blob/5c28b5b5eec9a4c1fb58541ed64d8f1fa3a99983/LZ4/src/H5Zlz4.c#L156

with the error:

    decompressed size not the same: -81, != 96

ajelenak commented:
Are you okay with this capitalization change @byrnHDF ? I wanted to ask you but you responded faster.

byrnHDF commented:
Was there a reason - because it will have consequences in other projects.

Also is it consistent with the other filters - or will those need changed as well.

Allen

________________________________________
From: Aleksandar Jelenak ***@***.***>
Sent: Wednesday, June 4, 2025 11:57 AM
To: HDFGroup/hdf5_plugins
Cc: Allen Byrne; Mention
Subject: Re: [HDFGroup/hdf5_plugins] Use better cd_values[] in LZ4 example (PR #186)

[https://avatars.githubusercontent.com/u/7267124?s=20&v=4]ajelenak left a comment (HDFGroup/hdf5_plugins#186)<https://github.com/HDFGroup/hdf5_plugins/pull/186#issuecomment-2940754136>

Are you okay with this capitalization change @byrnHDF<https://github.com/byrnHDF> ? I wanted to ask you but you responded faster.

—
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5_plugins/pull/186#issuecomment-2940754136>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AL77KBWKHCIVKWHZQBE6FAL3B4QPBAVCNFSM6AAAAAB6G5OHRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNBQG42TIMJTGY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ajelenak commented:
> Was there a reason - because it will have consequences in other projects.

Projects out there, or in-house, like HDF5?

The reason was the name of the filter is LZ4, not lz4. It is not a required change. Now that I better understand how many places need to be changed, I will switch back.

byrnHDF commented:
Yes, usually my choice too, if it involves a lot more work for little gain.

________________________________________
From: Aleksandar Jelenak ***@***.***>
Sent: Wednesday, June 4, 2025 1:28 PM
To: HDFGroup/hdf5_plugins
Cc: Allen Byrne; Mention
Subject: Re: [HDFGroup/hdf5_plugins] Use better cd_values[] in LZ4 example (PR #186)

[https://avatars.githubusercontent.com/u/7267124?s=20&v=4]ajelenak left a comment (HDFGroup/hdf5_plugins#186)<https://github.com/HDFGroup/hdf5_plugins/pull/186#issuecomment-2940998184>

Was there a reason - because it will have consequences in other projects.

Projects out there, or in-house, like HDF5?

The reason was the name of the filter is LZ4, not lz4. It is not a required change. Now that I better understand how many places need to be changed, I will switch back.

—
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5_plugins/pull/186#issuecomment-2940998184>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AL77KBTPCHGXX23IF7VIEFL3B43D3AVCNFSM6AAAAAB6G5OHRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNBQHE4TQMJYGQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ajelenak commented:
Ok, I removed those changes. Back to the important stuff...

I switched to an older version of the LZ4 plugin in this branch, and it worked for a more sensible block size in the LZ4 example. Also all LZ4-related tests are updated accordingly. We now need to decide how to proceed. It would be nice to understand why introduction of the `LZ4_decompress_safe()` function in the LZ4 plugin is not working with a block size larger than 3 bytes but also it is important to have a really working LZ4 plugin in this repository.

ajelenak commented:
The LZ4 plugin seems to work now using the new `LZ4_decompress_safe()` function. The updated LZ4 example code does not report an error about decompressed block size difference.

nhz2 commented on LZ4/src/H5Zlz4.c:
```diff
@@ -143,19 +143,24 @@ static size_t H5Z_filter_lz4(unsigned int flags, size_t cd_nelmts,
+                    goto error;
+                }
 #else
 #if LZ4_VERSION_NUMBER > 10800
```
`LZ4_decompress_safe` existed before `LZ4_VERSION_NUMBER` existed https://github.com/lz4/lz4/commit/64547df2dec94632ca8d5dd42b07c22a169b70b7

So I think the version checks can be safely removed.

ajelenak replied:
Yes, it would simplify the code. I am not familiar with the LZ4 evolution, is there any chance that older versions of this function were slower than the current implementation?

nhz2 commented on LZ4/src/H5Zlz4.c:
```diff
@@ -143,19 +143,24 @@ static size_t H5Z_filter_lz4(unsigned int flags, size_t cd_nelmts,
             {
 #if LZ4_VERSION_NUMBER > 10900
-                int compressedBytes = LZ4_decompress_safe(rpos, roBuf, blockSize, origSize);
+                int compressedBytes = LZ4_decompress_safe(rpos, roBuf, compressedBlockSize, origSize);
+                if(compressedBytes != blockSize)
+                {
+                    printf("decompressed size not the same: %d, != %d\n", compressedBytes, blockSize);
+                    goto error;
+                }
```
```suggestion
                int decompressedBytes = LZ4_decompress_safe(rpos, roBuf, compressedBlockSize, blockSize);
                if(decompressedBytes != blockSize)
                {
                    printf("decompressed size not the same: %d, != %d\n", decompressedBytes, blockSize);
                    goto error;
                }
```

From the lz4.h file, the last argument of `LZ4_decompress_safe` https://github.com/lz4/lz4/blob/dev/lib/lz4.h:

> dstCapacity : is the size of destination buffer (which must be already allocated), presumed an upper bound of decompressed size.

So it should be `blockSize` not `origSize`.

ajelenak replied:
Right, don't know how I missed this. Changed.
